// isaacscript/dist/src/checkForWindowsTerminalBugs.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const chalk_1 = __importDefault(require("chalk"));
const fs_1 = __importDefault(require("fs"));
const os_1 = __importDefault(require("os"));
const path_1 = __importDefault(require("path"));
const prompts_1 = __importDefault(require("prompts"));
const file = __importStar(require("./file"));
const misc_1 = require("./misc");
// https://stackoverflow.com/questions/9080085/node-js-find-home-directory-in-platform-agnostic-way
const homeDir = os_1.default.homedir();
const BASH_PROFILE_PATH = path_1.default.join(homeDir, ".bash_profile");
const WINDOWS_CODE_PAGE = "65001";
// By default, Git Bash for Windows uses MINGW64
// This will not work correctly with the prompts library (or any other NodeJS input library)
// Try to detect this and warn the end-user
async function checkForWindowsTerminalBugs() {
    if (process.platform !== "win32") {
        return;
    }
    if (process.env.SHELL !== "C:\\Program Files\\Git\\usr\\bin\\bash.exe") {
        return;
    }
    await checkForCodePage();
    await checkForWindowsBugColor();
}
exports.default = checkForWindowsTerminalBugs;
async function checkForCodePage() {
    const chcpPath = "C:\\Windows\\System32\\chcp.com";
    const stdout = misc_1.execExe(chcpPath).trim();
    // The output of "chcp" will be different depending on the language of the Windows installation,
    // so we only parse the final number
    const match = /(\d+)$/.exec(stdout);
    if (match === null) {
        misc_1.error(`Failed to parse the results of "${chalk_1.default.green(chcpPath)}":`, stdout);
    }
    const activeCodePage = match[1].trim();
    if (activeCodePage === WINDOWS_CODE_PAGE) {
        return;
    }
    console.error(chalk_1.default.red(`Error: It looks like you are using Git Bash for Windows (MINGW64) with an incorrect code page of "${chalk_1.default.green(activeCodePage)}". (It should be equal to "${WINDOWS_CODE_PAGE}".)`));
    const response = await prompts_1.default({
        type: "confirm",
        name: "fixCodePage",
        message: 'Do you want me to fix this for you? (Type "y", then press "enter".)',
        initial: true,
    });
    if (response.fixCodePage === false) {
        misc_1.error("Ok then. Good-bye.");
    }
    applyFixesToBashProfile();
}
async function checkForWindowsBugColor() {
    if (process.env.FORCE_COLOR === "true") {
        return;
    }
    console.error(chalk_1.default.red(`Error: It looks like you are using Git Bash for Windows (MINGW64) without the "${chalk_1.default.green("FORCE_COLOR")}" environment variable.`));
    console.error("This is necessary in order for colors to work properly in the terminal.");
    const response = await prompts_1.default({
        type: "confirm",
        name: "fixColors",
        message: 'Do you want me to fix this for you? (Type "y", then press "enter".)',
        initial: true,
    });
    if (response.fixColors === false) {
        misc_1.error("Ok then. Good-bye.");
    }
    applyFixesToBashProfile();
}
function applyFixesToBashProfile() {
    // Check to see if the Bash profile has data
    let bashProfileContents;
    if (file.exists(BASH_PROFILE_PATH)) {
        bashProfileContents = file.read(BASH_PROFILE_PATH);
    }
    else {
        bashProfileContents = "";
    }
    // Prepare the text to append
    let newText = "";
    if (bashProfileContents !== "" &&
        bashProfileContents[bashProfileContents.length - 1] !== "\n") {
        // If the Bash profile exists and has data, it should end in a newline
        // Add an extra newline if this is not the case
        newText += "\n";
    }
    if (bashProfileContents !== "") {
        // If the Bash profile exists and has data, add an extra newline between the existing stuff and
        // the new lines added by us
        newText += "\n";
    }
    newText += "# Terminal fixes added by IsaacScript\n";
    newText += `chcp.com ${WINDOWS_CODE_PAGE} > /dev/null\n`;
    newText += "export FORCE_COLOR=true\n";
    try {
        fs_1.default.appendFileSync(BASH_PROFILE_PATH, newText);
    }
    catch (err) {
        misc_1.error(`Failed to append text to "${BASH_PROFILE_PATH}":`, err);
    }
    console.log(chalk_1.default.green("Complete!"), `I have added the terminal fixes to "${BASH_PROFILE_PATH}".`);
    console.log(chalk_1.default.red("Please close and re-open your terminal, then run this program again."));
    process.exit(0);
}
