// isaacscript/dist/src/constants.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PUBLISH_POST_COPY_PY_PATH = exports.PUBLISH_PRE_COPY_PY_PATH = exports.VERSION_TXT_PATH = exports.METADATA_XML_PATH = exports.MAIN_LUA = exports.MOD_SOURCE_PATH = exports.CONSTANTS_TS_PATH = exports.PACKAGE_JSON_PATH = exports.TSCONFIG_PATH = exports.CONFIG_FILE_PATH = exports.CONFIG_FILE_NAME = exports.README_MD_TEMPLATES_PATH = exports.README_MD = exports.PACKAGE_JSON_TEMPLATE_PATH = exports.PACKAGE_JSON = exports.METADATA_VDF_TEMPLATE_PATH = exports.METADATA_VDF = exports.METADATA_XML_TEMPLATE_PATH = exports.METADATA_XML = exports.MAIN_TS_TEMPLATE_PATH = exports.MAIN_TS = exports.GITIGNORE_TEMPLATE_PATH = exports.GITIGNORE = exports.TEMPLATES_STATIC_DIR = exports.WATCHER_MOD_SOURCE_PATH = exports.WATCHER_MOD_NAME = exports.DISABLE_IT_FILE = exports.FILE_SYNCED_MESSAGE = exports.CWD = exports.CURRENT_DIRECTORY_NAME = void 0;
const path_1 = __importDefault(require("path"));
const cwd = process.cwd();
// Miscellaneous
exports.CURRENT_DIRECTORY_NAME = path_1.default.basename(cwd);
exports.CWD = cwd;
exports.FILE_SYNCED_MESSAGE = "File synced:";
// isaacscript
const REPO_ROOT = path_1.default.join(__dirname, "..", "..");
// isaacscript/isaacscript-watcher
exports.DISABLE_IT_FILE = "disable.it";
exports.WATCHER_MOD_NAME = "isaacscript-watcher";
exports.WATCHER_MOD_SOURCE_PATH = path_1.default.join(REPO_ROOT, exports.WATCHER_MOD_NAME);
// isaacscript/file-templates
const TEMPLATES_DIR = path_1.default.join(REPO_ROOT, "file-templates");
// isaacscript/file-templates/static
exports.TEMPLATES_STATIC_DIR = path_1.default.join(TEMPLATES_DIR, "static");
// isaacscript/file-templates/dynamic
const TEMPLATES_DYNAMIC_DIR = path_1.default.join(TEMPLATES_DIR, "dynamic");
exports.GITIGNORE = "gitignore"; // Not named ".gitignore" to prevent NPM from deleting it
exports.GITIGNORE_TEMPLATE_PATH = path_1.default.join(TEMPLATES_DYNAMIC_DIR, exports.GITIGNORE);
exports.MAIN_TS = "main.ts";
exports.MAIN_TS_TEMPLATE_PATH = path_1.default.join(TEMPLATES_DYNAMIC_DIR, exports.MAIN_TS);
exports.METADATA_XML = "metadata.xml";
exports.METADATA_XML_TEMPLATE_PATH = path_1.default.join(TEMPLATES_DYNAMIC_DIR, exports.METADATA_XML);
exports.METADATA_VDF = "metadata.vdf";
exports.METADATA_VDF_TEMPLATE_PATH = path_1.default.join(TEMPLATES_DYNAMIC_DIR, exports.METADATA_VDF);
exports.PACKAGE_JSON = "package.json";
exports.PACKAGE_JSON_TEMPLATE_PATH = path_1.default.join(TEMPLATES_DYNAMIC_DIR, exports.PACKAGE_JSON);
exports.README_MD = "README.md";
exports.README_MD_TEMPLATES_PATH = path_1.default.join(TEMPLATES_DYNAMIC_DIR, exports.README_MD);
// project
exports.CONFIG_FILE_NAME = "isaacscript.json";
exports.CONFIG_FILE_PATH = path_1.default.join(exports.CWD, exports.CONFIG_FILE_NAME);
exports.TSCONFIG_PATH = path_1.default.join(exports.CWD, "tsconfig.json");
exports.PACKAGE_JSON_PATH = path_1.default.join(exports.CWD, "package.json");
exports.CONSTANTS_TS_PATH = path_1.default.join(exports.CWD, "src", "constants.ts");
// project/mod
exports.MOD_SOURCE_PATH = path_1.default.join(exports.CWD, "mod");
exports.MAIN_LUA = "main.lua";
exports.METADATA_XML_PATH = path_1.default.join(exports.MOD_SOURCE_PATH, "metadata.xml");
exports.VERSION_TXT_PATH = path_1.default.join(exports.MOD_SOURCE_PATH, "version.txt");
// project/scripts
const PUBLISH_PRE_COPY_PY = "publish_pre_copy.py";
exports.PUBLISH_PRE_COPY_PY_PATH = path_1.default.join(exports.CWD, "scripts", PUBLISH_PRE_COPY_PY);
const PUBLISH_POST_COPY_PY = "publish_post_copy.py";
exports.PUBLISH_POST_COPY_PY_PATH = path_1.default.join(exports.CWD, "scripts", PUBLISH_POST_COPY_PY);
