// isaacscript/dist/src/misc.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.snakeKebabToCamel = exports.parseIntSafe = exports.hasWhiteSpace = exports.getTime = exports.execShell = exports.execExe = exports.error = exports.ensureAllCases = void 0;
const chalk_1 = __importDefault(require("chalk"));
const child_process_1 = require("child_process");
const moment_1 = __importDefault(require("moment"));
const constants_1 = require("./constants");
// Use this on a switch statement's default case to get
// the linter to complain if a case was not predicted
const ensureAllCases = (obj) => obj;
exports.ensureAllCases = ensureAllCases;
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function error(...args) {
    console.error(...args);
    process.exit(1);
}
exports.error = error;
function execExe(path, cwd = constants_1.CWD) {
    let stdout;
    try {
        const buffer = child_process_1.execSync(`"${path}"`, {
            cwd,
        });
        stdout = buffer.toString().trim();
    }
    catch (err) {
        error(`Failed to run "${chalk_1.default.green(path)}":`, err);
    }
    return stdout;
}
exports.execExe = execExe;
function execShell(command, args, allowFailure = false, cwd = constants_1.CWD) {
    // On Windows, "spawnSync()" will not account for spaces in arguments
    // Thus, wrap everything in a double quote
    // This will cause arguments that naturally have double quotes to fail
    for (let i = 0; i < args.length; i++) {
        args[i] = `"${args[i]}"`; // eslint-disable-line no-param-reassign
    }
    const commandDescription = `${command} ${args.join(" ")}`;
    let spawnSyncReturns;
    try {
        spawnSyncReturns = child_process_1.spawnSync(command, args, {
            shell: true,
            cwd,
        });
    }
    catch (err) {
        error(`Failed to run the "${chalk_1.default.green(commandDescription)}" command:`, err);
    }
    const exitStatus = spawnSyncReturns.status;
    const stdout = spawnSyncReturns.output.join("\n").trim();
    if (exitStatus !== 0) {
        if (allowFailure) {
            return [exitStatus, stdout];
        }
        console.error(`Failed to run the "${chalk_1.default.green(commandDescription)}" command with an exit code of ${exitStatus}.`);
        console.error("The output was as follows:");
        console.error(stdout);
        process.exit(1);
    }
    return [exitStatus, stdout];
}
exports.execShell = execShell;
function getTime() {
    return moment_1.default().format("h:mm:ss A"); // e.g. "1:23:45 AM"
}
exports.getTime = getTime;
// From: https://stackoverflow.com/questions/1731190/check-if-a-string-has-white-space
function hasWhiteSpace(s) {
    return /\s/g.test(s);
}
exports.hasWhiteSpace = hasWhiteSpace;
// parseIntSafe is a more reliable version of parseInt
// By default, "parseInt('1a')" will return "1", which is unexpected
// This returns either an integer or NaN
function parseIntSafe(input) {
    if (typeof input !== "string") {
        return NaN;
    }
    // Remove all leading and trailing whitespace
    let trimmedInput = input.trim();
    const isNegativeNumber = trimmedInput.startsWith("-");
    if (isNegativeNumber) {
        // Remove the leading minus sign before we match the regular expression
        trimmedInput = trimmedInput.substring(1);
    }
    if (/^\d+$/.exec(trimmedInput) === null) {
        // "\d" matches any digit (same as "[0-9]")
        return NaN;
    }
    if (isNegativeNumber) {
        // Add the leading minus sign back
        trimmedInput = `-${trimmedInput}`;
    }
    return parseInt(trimmedInput, 10);
}
exports.parseIntSafe = parseIntSafe;
// Convert snake_case and kebab-case to camelCase
// From: https://hisk.io/javascript-snake-to-camel/
function snakeKebabToCamel(str) {
    return str.replace(/([-_][a-z])/g, (group) => group.toUpperCase().replace("-", "").replace("_", ""));
}
exports.snakeKebabToCamel = snakeKebabToCamel;
