// isaacscript/dist/src/commands/init/createMod.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = __importDefault(require("path"));
const configFile = __importStar(require("../../configFile"));
const constants_1 = require("../../constants");
const file = __importStar(require("../../file"));
const misc_1 = require("../../misc");
function createMod(projectName, projectPath, createNewDir, modsDirectory, saveSlot) {
    if (createNewDir) {
        file.makeDir(projectPath);
    }
    const config = configFile.createObject(modsDirectory, saveSlot);
    configFile.createFile(projectPath, config);
    const targetModDirectory = path_1.default.join(config.modsDirectory, projectName);
    makeSubdirectories(projectPath);
    copyStaticFiles(projectPath);
    copyDynamicFiles(projectName, projectPath, targetModDirectory);
    updateNodeModules(projectPath);
    installNodeModules(projectPath);
}
exports.default = createMod;
function makeSubdirectories(projectPath) {
    for (const subdirectory of ["mod", "src"]) {
        const srcPath = path_1.default.join(projectPath, subdirectory);
        file.makeDir(srcPath);
    }
}
// Copy static files, like ".eslintrc.js", "tsconfig.json", etc.
function copyStaticFiles(projectPath) {
    const staticFileList = file.getDirList(constants_1.TEMPLATES_STATIC_DIR);
    staticFileList.forEach((fileName) => {
        const templateFilePath = path_1.default.join(constants_1.TEMPLATES_STATIC_DIR, fileName);
        const destinationFilePath = path_1.default.join(projectPath, fileName);
        if (!file.exists(destinationFilePath)) {
            file.copy(templateFilePath, destinationFilePath);
        }
    });
}
// Copy files that need to have text replaced inside of them
function copyDynamicFiles(projectName, projectPath, targetModDirectory) {
    // ".gitignore"
    {
        const fileName = constants_1.GITIGNORE;
        const templatePath = constants_1.GITIGNORE_TEMPLATE_PATH;
        const template = file.read(templatePath);
        // Prepend a header with the project name
        let separatorLine = "# ";
        for (let i = 0; i < projectName.length; i++) {
            separatorLine += "-";
        }
        separatorLine += "\n";
        const gitignoreHeader = `${separatorLine}# ${projectName}\n${separatorLine}\n`;
        const gitignore = gitignoreHeader + template;
        const destinationPath = path_1.default.join(projectPath, `.${fileName}`); // We need to prepend a period
        file.write(destinationPath, gitignore);
    }
    // "package.json"
    {
        // Modify and copy the file
        const fileName = constants_1.PACKAGE_JSON;
        const templatePath = constants_1.PACKAGE_JSON_TEMPLATE_PATH;
        const template = file.read(templatePath);
        const packageJSON = template.replace(/MOD_NAME/g, projectName);
        const destinationPath = path_1.default.join(projectPath, fileName);
        file.write(destinationPath, packageJSON);
    }
    // "README.md"
    {
        const fileName = constants_1.README_MD;
        const templatePath = constants_1.README_MD_TEMPLATES_PATH;
        const template = file.read(templatePath);
        const readmeMD = template.replace(/MOD_NAME/g, projectName);
        const destinationPath = path_1.default.join(projectPath, fileName);
        file.write(destinationPath, readmeMD);
    }
    // "mod/metadata.xml"
    {
        const fileName = constants_1.METADATA_XML;
        const templatePath = constants_1.METADATA_XML_TEMPLATE_PATH;
        const template = file.read(templatePath);
        const metadataXML = template.replace(/MOD_NAME/g, projectName);
        const modPath = path_1.default.join(projectPath, "mod");
        const destinationPath = path_1.default.join(modPath, fileName);
        file.write(destinationPath, metadataXML);
    }
    // "mod/metadata.vdf"
    {
        const fileName = constants_1.METADATA_VDF;
        const templatePath = constants_1.METADATA_VDF_TEMPLATE_PATH;
        const template = file.read(templatePath);
        const metadataVDF = template.replace(/MOD_TARGET_DIR/g, targetModDirectory);
        const modPath = path_1.default.join(projectPath, "mod");
        const destinationPath = path_1.default.join(modPath, fileName);
        file.write(destinationPath, metadataVDF);
    }
    // "src/main.ts"
    {
        // Convert snake_case and kebab-case to camelCase
        // (kebab-case in particular will make the example TypeScript file fail to compile)
        const srcPath = path_1.default.join(projectPath, "src");
        const camelCaseProjectName = misc_1.snakeKebabToCamel(projectName);
        const fileName = constants_1.MAIN_TS;
        const templatePath = constants_1.MAIN_TS_TEMPLATE_PATH;
        const template = file.read(templatePath);
        const mainTS = template
            .replace(/MOD_NAME initialized/g, `${projectName} initialized`)
            .replace(/MOD_NAME/g, camelCaseProjectName);
        const destinationPath = path_1.default.join(srcPath, fileName);
        file.write(destinationPath, mainTS);
    }
}
function updateNodeModules(projectPath) {
    console.log("Updating node modules...");
    misc_1.execShell("npx", ["npm-check-updates", "--upgrade", "--packageFile", "package.json"], false, projectPath);
}
function installNodeModules(projectPath) {
    console.log("Installing node modules... (This can take a long time.)");
    misc_1.execShell("npm", ["install"], false, projectPath);
}
